### ISSUE TITLE
Check failure errors in grpcio==1.68.1

### ISSUE THREAD

**Issue by kramarz** [grpc/grpc#38251]
### What version of gRPC and what language are you using?
Python grpcio==1.68.1

### What operating system (Linux, Windows,...) and version?
Cloud Run

### What runtime / compiler are you using (e.g. python version or version of gcc)

base docker image: python:3.10-slim

### What did you do?

using google_cloud_spanner library suddenly started causing transient errors after releasing a new version of a service. After troubleshooting we realized that it was caused by installing the latest version of grpcio python library.

### What did you expect to see?

Downgrading grpcio to 1.67.1 fixed the issue.

### What did you see instead?

Transient errors:

```
DEFAULT 2024-11-27T13:14:28.517509Z F0000 00:00:1732713268.514118 2817 thd.h:170] Check failed: state_ == FAILED
DEFAULT 2024-11-27T13:14:28.517797Z *** Check failure stack trace: ***
ERROR 2024-11-27T13:14:28.517917402Z Uncaught signal: 6, pid=1, tid=2817, fault_addr=0.
WARNING 2024-11-27T13:14:29.298014698Z Container terminated on signal 6.
```

Or

```
DEFAULT 2024-12-02T18:51:56.088031Z F0000 00:00:1733165516.085607 469 work_stealing_thread_pool.cc:239] Check failed: !IsQuiesced()
DEFAULT 2024-12-02T18:51:56.093679Z *** Check failure stack trace: ***
ERROR 2024-12-02T18:51:56.093710906Z Uncaught signal: 6, pid=469, tid=469, fault_addr=0.
```

### Anything else we should know about your project / environment?

I am reporting this as GCP Support on behalf of a customer.

**Comment by sourabhsinghs** [comment]
@eugeneo  Can you take a look into this once

**Comment by imrehg** [comment]
We've seen similar transient errors on AWS/Fargate with containers failing ~30% of the time due to this. Version 1.67.1 works fine.

**Comment by eugeneo** [comment]
@drfloob Any ideas?

**Comment by jrp-enf** [comment]
I've seen the "thd.h:170] Check failed: state_ == FAILED" in a radically different context, also transient.

Will try to drop down to 1.67.1.

**Comment by villainb-dg** [comment]
FYI: we've also seen this error on v1.65.5:

```
thd.h:172] Check failed: state_ == FAILED
```

**Comment by atollena** [comment]
As reported by @villainb-dg we (datadog) are running into this issue. It happens often enough that we could probably capture more info (core dump, grpc trace logs, etc.). Is there anything in particular we could do to help the investigation?

**Comment by villainb-dg** [comment]
We do have a reproduction with `timer_check` logs on. They don't provide a lot of insights given that it's hard to tell which of the parent or child process they come from (this occurs during a fork in our case). Looking at the code, I was able to identify (I think) a use-after-free bug that could explain this behavior. The [gc_completed_threads](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/timer_manager.cc#L67-L80) method acquires the lock after assigning a thread local variable `to_gc`. If two threads are executing this function concurrently, it's possible that both attempt to call `Join` which is illegal as per the `Thread` class definition (see https://github.com/grpc/grpc/blob/84d143b0045ac60f273a9f5e9c6588207a73bdca/src/core/util/thd.h#L163-L172).

**Comment by villainb-dg** [comment]
The root cause of this issue seems to be a race condition in the `grpc_core::Executor` class. It can cause this assertion to fail but I was also able to trigger a deadlock on the parent process. The `Enqueue` [method ](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/executor.cc#L331-L352) creates a new `Thread` holding the `adding_thread_lock_` lock but the decision to create the thread is done solely holding the thread lock:

```
      try_new_thread = ts->depth > MAX_DEPTH &&
                       cur_thread_count < max_threads_ && !ts->shutdown;

      ts->queued_long_job = !is_short;

      gpr_mu_unlock(&ts->mu);
      break;
    }

    if (try_new_thread && gpr_spinlock_trylock(&adding_thread_lock_)) {
      cur_thread_count = static_cast<size_t>(gpr_atm_acq_load(&num_threads_));
      if (cur_thread_count < max_threads_) {
        // Increment num_threads (safe to do a store instead of a cas because we
        // always increment num_threads under the 'adding_thread_lock')
        gpr_atm_rel_store(&num_threads_, cur_thread_count + 1);

        thd_state_[cur_thread_count].thd =
            Thread(name_, &Executor::ThreadMain, &thd_state_[cur_thread_count]);
        thd_state_[cur_thread_count].thd.Start();
      }
      gpr_spinlock_unlock(&adding_thread_lock_);
    }
```

This means that during the `SetThreading(false)` call, the statement setting the thread state to `shutdown` can be executed after the decision to create a new thread is made:

```
    for (size_t i = 0; i < max_threads_; i++) {
      gpr_mu_lock(&thd_state_[i].mu);
      thd_state_[i].shutdown = true;
      gpr_cv_signal(&thd_state_[i].cv);
      gpr_mu_unlock(&thd_state_[i].mu);
    }
```

The problem is that the following statement makes the wrong assumption that acquiring the `adding_thread_lock_` is going to be enough to guarantee no new thread is created:

```
    // Ensure no thread is adding a new thread. Once this is past, then no
    // thread will try to add a new one either (since shutdown is true)
    gpr_spinlock_lock(&adding_thread_lock_);
    gpr_spinlock_unlock(&adding_thread_lock_);

    curr_num_threads = gpr_atm_no_barrier_load(&num_threads_);
    for (gpr_atm i = 0; i < curr_num_threads; i++) {
      thd_state_[i].thd.Join();
      GRPC_TRACE_LOG(executor, INFO)
          << "EXECUTOR (" << name_ << ") Thread " << i + 1 << " of "
          << curr_num_threads << " joined";
    }
```

Reading `num_threads_` on line 165 may occur after the create thread routine has increased it (line 345) but before the concrete thread was actually assigned in the `thd_state_` array. If the `Join` is done before the thread is created, then it would trigger the assertion to fail on thd.h (since dummy ones are created):

```
    for (size_t i = 0; i < max_threads_; i++) {
      gpr_mu_init(&thd_state_[i].mu);
      gpr_cv_init(&thd_state_[i].cv);
      thd_state_[i].id = i;
      thd_state_[i].name = name_;
      thd_state_[i].thd = Thread();
      thd_state_[i].elems = GRPC_CLOSURE_LIST_INIT;
    }
```

I think the right fix is to validate that the thread state wasn't set to `shutdown` before attempting the thread creation (line 340). The thread lock needs to be held but that should be fine performance-wise.

**Comment by sreenithi** [comment]
@villainb-dg thanks for this detailed investigation. I see that you also have a commit with the fix. Would you be willing to raise a PR with the fix?
@eugeneo @drfloob please take a look at this and share your comments too

**Comment by villainb-dg** [comment]
Yes I'm in the process of trying to reproduce it within the unit test framework of gRPC. I wanted to add tests to attempt at catching any regression. The issue might be that it may catch other deadlocks as I was able to find and document more (will follow-up).

**Comment by drfloob** [comment]
Since it isn't mentioned in this thread, I'll not this hang occurs on `fork`, likely implying some kind of pre-existing bug in iomgr's fork implementation. The Executor class is only used in Python and Ruby now, and will be deleted when @eugeneo has finished the new fork support code, so any fixes we do in Executor will be fairly short-lived.

@villainb-dg can you confirm that the problem still persists in v1.71.0?

**Comment by drfloob** [comment]
I've commented on https://github.com/grpc/grpc/pull/38924, and assigned it to the Python team. If the Python fork tests are not sensitive to a real bug, then they won't be sensitive to a regression either.

**Comment by jc21** [comment]
Hi, just want to chime in and say I'm seeing this in our PHP GRPC extension too. It's happening repeatedly for us when running PHPUnit test suites in our CI. However, I cannot replicate this when running the tests locally.

```
php -i | grep -i grpc
grpc support => enabled
grpc module version => 1.70.0
grpc.enable_fork_support => 1 => 1
```

```
F0000 00:00:1741845039.466574   72176 work_stealing_thread_pool.cc:239] Check failed: !IsQuiesced() 
*** Check failure stack trace: ***
```

Looking forward to seeing if the PR fixes it.

**Comment by villainb-dg** [comment]
> The Executor class is only used in Python and Ruby now, and will be deleted when eugeneo has finished the new fork support code, so any fixes we do in Executor will be fairly short-lived.

@drfloob any timeline you can share for this work to be released? Cheers!

**Comment by drfloob** [comment]
I'm not particularly relevant here, please ping @eugeneo and @sreenithi going forward.

**Comment by murgatroid99** [comment]
If this bug is in fact related to how iomgr handles fork, it may be resolved by #41432, which changed the Python and Ruby libraries to use EventEngine instead of iomgr to handle fork. That change was published in version 1.80.

**Comment by sergiitk** [comment]
Can anyone confirm this is fixed or share repro code so we can test it ourselves?

**Comment by sergiitk** [comment]
No response to provide more information, so closing. More information can be provided and then we can reopen.
